fms/utils/serialization.py

```diff
@@ -38,18 +71,28 @@ def register_adapter(
     architecture: The name of the model architecture, e.g. 'llama'
     source: A label representing the format of the weights to be converted.
             E.g. 'hf'
-    adapter: the class of the adapter. The class must accept one constructor
-                parameter, which will be a state dict (`OrderedDict`)
+    adapter_steps: a list of registered steps to build the basic adapter for an
+            architecture and source combination. This can be augmented with extra
+            steps if needed during call to _get_adapter()
     """
-    sources: MutableMapping[str, Callable[[Mapping], Mapping]] = {}
+    sources: MutableMapping[
+        str, Callable[[Mapping[str, Any], Mapping[str, Any]], Mapping[str, Any]]
+    ] = {}
     if architecture in __adapters:
         sources = __adapters[architecture]
 
     if source in sources:
         raise KeyError(
-            f"Variant {source} already registered for architecture {architecture}"
+            f"Source {source} already registered for architecture {architecture}"
         )
 
+    # Create a new base adapter for this source
+    step_functions = [__adapter_steps[architecture][step] for step in adapter_steps]
+    adapter = lambda initial_sd, extra_args: reduce(
+        lambda state_dict, step_func: step_func(state_dict, extra_args),
```

Comment:
Just a thought, but rather than extra_args being a mapping, could we pass this in as **extra_args, that way a user can create their adapter step function with:

```python
def my_func(sd, model_config=None)
```